File: algorithmic_toolbox/week2_algorithmic_warmup/3_greatest_common_divisor/gcd.py
import sys


def gcd_naive(a: int, b: int) -> int:  # noqa: C901
    current_gcd = 1

    for d in range(2, min(a, b) + 1):
        if a % d == b % d == 0:  # noqa: SIM102
            if d > current_gcd:
                current_gcd = d

    return current_gcd


def gcd_euclidean(a: int, b: int) -> int:
    _min = min(a, b)
    _max = max(a, b)

    return _max if _min == 0 else gcd_euclidean(_max % _min, _min)


if __name__ == "__main__":
    a, b = map(int, sys.stdin.read().split())
    print(gcd_naive(a, b))
    print(gcd_euclidean(a, b))

File: algorithmic_toolbox/week2_algorithmic_warmup/3_greatest_common_divisor/gcd.cpp
#include <algorithm>
#include <iostream>

int gcd_naive(int a, int b) {
  int current_gcd = 1;

  for (int d = 2; d <= a && d <= b; d++)
    if (a % d == 0 && b % d == 0)
      if (d > current_gcd)
        current_gcd = d;

  return current_gcd;
}

int gcd_euclidean(int a, int b) {
  int min = std::min(a, b);
  int max = std::max(a, b);

  return min == 0 ? max : gcd_euclidean(max % min, min);
}

int main() {
  int a, b;
  std::cin >> a >> b;

  std::cout << gcd_naive(a, b) << std::endl;
  std::cout << gcd_euclidean(a, b) << std::endl;

  return 0;
}
